src/DataSet/AttributeDataSet.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\DataSet;
+
+use InvalidArgumentException;
+use ReflectionClass;
+use ReflectionException;
+use Yiisoft\Validator\Attribute\HasMany;
+use Yiisoft\Validator\Attribute\HasOne;
+use Yiisoft\Validator\Rule\Each;
+use Yiisoft\Validator\Rule\Nested;
+use Yiisoft\Validator\RuleInterface;
+use Yiisoft\Validator\RulesProviderInterface;
+
+use function in_array;
+
+/**
+ * This data set makes use of attributes introduced in PHP 8. It simplifies rules configuration process, especially for
+ * nested data and relations. Please refer to the guide for example.
+ *
+ * @link https://www.php.net/manual/en/language.attributes.overview.php
+ */
+final class AttributeDataSet implements RulesProviderInterface
+{
+    use ArrayDataTrait;
+
+    private object $baseAnnotatedObject;
+
+    public function __construct(object $baseAnnotatedObject, array $data = [])
+    {
+        $this->baseAnnotatedObject = $baseAnnotatedObject;
+        $this->data = $data;
+    }
+
+    public function getRules(): iterable
+    {
+        $classMeta = new ReflectionClass($this->baseAnnotatedObject);
+
+        return $this->handleAttributes($classMeta);
```

Comment:
What about traits support?

Reply:
I don't think they will be used directly. Could you provide an example?

Reply:
:white_check_mark: